File: examples/python-tool/frame.py
"""Length-prefixed JSON framing for the Aivyx tool process protocol.

Wire format (per docs/DAEMON_IPC.md):

    +-----------------+-----------------------------------+
    | 4 bytes         | N bytes                           |
    | big-endian u32  | UTF-8 JSON payload                |
    +-----------------+-----------------------------------+

This is a copy of `examples/python-channel/frame.py` — same wire
format, different transport (stdin/stdout instead of Unix socket).
Vendored rather than shared so each example stays independently
runnable.

Stdlib only.
"""

from __future__ import annotations

import io
import json
import struct
import sys
from typing import Any, BinaryIO

MAX_PAYLOAD_SIZE = 16 * 1024 * 1024
FRAME_HEADER_LEN = 4


class FrameError(Exception):
    """Malformed frame on the wire — bail out."""


def encode_frame(message: dict[str, Any]) -> bytes:
    body = json.dumps(message, separators=(",", ":")).encode("utf-8")
    if len(body) > MAX_PAYLOAD_SIZE:
        raise FrameError(f"payload {len(body)} bytes exceeds 16 MiB cap")
    header = struct.pack(">I", len(body))
    return header + body


def write_frame(stream: BinaryIO, message: dict[str, Any]) -> None:
    """Encode `message` and flush it to a binary stream."""
    stream.write(encode_frame(message))
    stream.flush()


def read_frame(stream: BinaryIO) -> dict[str, Any] | None:
    """Block until one full frame is available, decode it, return
    the parsed dict. Returns None on clean EOF (no header bytes
    were read before the stream closed).
    """
    header = _recv_exact(stream, FRAME_HEADER_LEN, allow_eof=True)
    if header is None:
        return None
    (length,) = struct.unpack(">I", header)
    if length > MAX_PAYLOAD_SIZE:
        raise FrameError(f"frame length {length} exceeds 16 MiB cap")
    body = _recv_exact(stream, length, allow_eof=False)
    try:
        return json.loads(body.decode("utf-8"))
    except (UnicodeDecodeError, json.JSONDecodeError) as exc:
        raise FrameError(f"invalid frame body: {exc}") from exc


def _recv_exact(stream: BinaryIO, n: int, allow_eof: bool) -> bytes | None:
    chunks: list[bytes] = []
    remaining = n
    while remaining > 0:
        chunk = stream.read(remaining)
        if not chunk:
            if allow_eof and not chunks:
                return None
            raise FrameError(
                f"EOF in frame after {n - remaining} of {n} bytes"
            )
        chunks.append(chunk)
        remaining -= len(chunk)
    return b"".join(chunks)

File: crates/aivyx-tool/src/frame.rs
//! Length-prefixed JSON framing for the tool process protocol.
//!
//! Wire format (per `docs/DAEMON_IPC.md`):
//!
//! ```text
//! +-----------------+-----------------------------------+
//! | 4 bytes         | N bytes                           |
//! | big-endian u32  | UTF-8 JSON payload                |
//! +-----------------+-----------------------------------+
//! ```
//!
//! Identical framing to `aivyx-channel::daemon_ipc`. Reimplemented
//! here so `aivyx-tool` does not depend on `aivyx-channel` (the
//! crate-graph edge would be the wrong direction — daemon side
//! pulls `aivyx-tool` in, not vice versa).

use serde::Serialize;
use thiserror::Error;
use tokio::io::{AsyncReadExt, AsyncWrite, AsyncWriteExt};

pub const MAX_PAYLOAD_SIZE: u32 = 16 * 1024 * 1024;
pub const FRAME_HEADER_LEN: usize = 4;

#[derive(Debug, Error)]
pub enum FrameError {
    #[error("payload size {0} bytes exceeds 16 MiB cap")]
    PayloadTooLarge(usize),
    #[error("frame length prefix {0} exceeds 16 MiB cap")]
    HeaderTooLarge(u32),
    #[error("serialize failed: {0}")]
    Serialize(String),
    #[error("io error: {0}")]
    Io(#[from] std::io::Error),
    #[error("invalid utf-8 in frame body: {0}")]
    Utf8(#[from] std::string::FromUtf8Error),
    #[error("invalid json in frame body: {0}")]
    Json(#[from] serde_json::Error),
}

/// Encode `message` as a length-prefixed frame. Returns the bytes
/// to write to the wire.
pub fn encode_frame<T: Serialize>(message: &T) -> Result<Vec<u8>, FrameError> {
    let body =
        serde_json::to_vec(message).map_err(|e| FrameError::Serialize(e.to_string()))?;
    if body.len() > MAX_PAYLOAD_SIZE as usize {
        return Err(FrameError::PayloadTooLarge(body.len()));
    }
    let len = body.len() as u32;
    let mut out = Vec::with_capacity(FRAME_HEADER_LEN + body.len());
    out.extend_from_slice(&len.to_be_bytes());
    out.extend_from_slice(&body);
    Ok(out)
}

/// Read one full frame from an async reader. Blocks until the
/// header + body are available. The returned bytes are the
/// decoded UTF-8 JSON body — the caller can `serde_json::from_str`
/// it into whatever shape they need.
///
/// Returns `Ok(None)` if the reader is at EOF before any header
/// bytes arrive (peer closed cleanly). Returns `Err` on a partial
/// frame.
pub async fn read_frame<R: AsyncReadExt + Unpin>(
    reader: &mut R,
) -> Result<Option<String>, FrameError> {
    let mut header = [0u8; FRAME_HEADER_LEN];
    let mut bytes_read = 0;
    while bytes_read < FRAME_HEADER_LEN {
        match reader.read(&mut header[bytes_read..]).await {
            Ok(0) => {
                if bytes_read == 0 {
                    return Ok(None); // clean EOF
                }
                return Err(FrameError::Io(std::io::Error::new(
                    std::io::ErrorKind::UnexpectedEof,
                    format!(
                        "EOF in frame header after {bytes_read} of {FRAME_HEADER_LEN} bytes"
                    ),
                )));
            }
            Ok(n) => bytes_read += n,
            Err(e) => return Err(FrameError::Io(e)),
        }
    }
    let len = u32::from_be_bytes(header);
    if len > MAX_PAYLOAD_SIZE {
        return Err(FrameError::HeaderTooLarge(len));
    }
    let mut body = vec![0u8; len as usize];
    reader.read_exact(&mut body).await?;
    let text = String::from_utf8(body)?;
    Ok(Some(text))
}

/// Write one frame to an async writer. Convenience wrapper —
/// combines `encode_frame` and `write_all`.
pub async fn write_frame<W: AsyncWrite + Unpin, T: Serialize>(
    writer: &mut W,
    message: &T,
) -> Result<(), FrameError> {
    let bytes = encode_frame(message)?;
    writer.write_all(&bytes).await?;
    writer.flush().await?;
    Ok(())
}

#[cfg(test)]
mod tests {
    use super::*;
    use tokio::io::duplex;

    #[derive(Debug, serde::Serialize, serde::Deserialize, PartialEq)]
    struct Sample {
        kind: String,
        value: i32,
    }

    #[tokio::test]
    async fn encode_then_decode_round_trips() {
        let (mut a, mut b) = duplex(8192);
        let sample = Sample {
            kind: "hello".into(),
            value: 42,
        };
        write_frame(&mut a, &sample).await.unwrap();
        let body = read_frame(&mut b).await.unwrap().expect("non-EOF");
        let back: Sample = serde_json::from_str(&body).unwrap();
        assert_eq!(back, sample);
    }

    #[tokio::test]
    async fn read_frame_returns_none_on_clean_eof() {
        let (a, mut b) = duplex(8192);
        drop(a); // close the write side cleanly
        let result = read_frame(&mut b).await.unwrap();
        assert!(result.is_none());
    }

    #[tokio::test]
    async fn read_frame_errors_on_partial_header() {
        let (mut a, mut b) = duplex(8192);
        a.write_all(&[0u8, 1u8]).await.unwrap();
        drop(a);
        let result = read_frame(&mut b).await;
        assert!(result.is_err());
    }

    #[tokio::test]
    async fn header_too_large_is_rejected() {
        let (mut a, mut b) = duplex(8192);
        let bogus_len = (MAX_PAYLOAD_SIZE + 1).to_be_bytes();
        a.write_all(&bogus_len).await.unwrap();
        drop(a);
        let result = read_frame(&mut b).await;
        match result {
            Err(FrameError::HeaderTooLarge(_)) => {}
            other => panic!("expected HeaderTooLarge, got {other:?}"),
        }
    }
}
